### ISSUE TITLE
Requirements/Design: Consumer Groups

### ISSUE THREAD

**Issue by nicholastmosher** [infinyon/fluvio#760]
A Consumer Group is an abstraction that allows coordination among several consumers that are working on the same logical task. It allows for multiple partitions to be read in parallel (increasing throughput) without causing any two consumers in the group to process the same message twice. In this issue, I'd like to put forward some initial requirements, including functional requirements, potential API requirements, etc.

At the highest level, a client should be able to do the following operations:

- Create a new consumer group
- Join an existing consumer group
- Receive messages from an assigned partition
- Manually or automatically "commit" a particular offset on a partition as having been consumed

The cluster should be able to do the following:

- Create new consumer groups for a topic
- Accept new consumer clients and add them to an existing consumer group
- Assign consumers to partitions
- Send messages from a partition to the appropriate assigned consumer
- Accept offset commits from consumers
- Detect when a consumer in a consumer group goes offline or otherwise becomes unresponsive
- Re-assign a partition to a new consumer (e.g. if the partition's primary consumer went offline)

Based on my understanding of consumer groups, it seems that the following statements should hold true:

- A consumer group must have exactly one associated topic.
- A topic may have zero or more consumer groups.
- A consumer group must have a unique identifier (used by consumer-clients to join the group)
- A consumer group must map all partitions in a topic to consumer-clients
- A partition must not be assigned to more than one consumer-client (prevent double-reads)
- A consumer group must maintain the currently-committed offset of each partition

Here are some additional thoughts, questions, and ideas I have had while thinking about consumer groups

- Can we assign human-readable names to consumer groups? This could allow us to create an elegant API that is easy for application developers. For example, when joining a consumer group, rather than needing to provide a numeric consumer-group-id, we could allow clients to provide a consumer-group-name.
- Could all consumers be considered a special case of a consumer group? An application with only one consumer-client could use a consumer group where all partitions are mapped onto the single consumer-client.
  - Potential downside: what if we do not want the overhead of remembering offsets on the cluster-side?
- Is it desirable to have both client-side offset management and cluster-side offset management? (related to previous point)
- Should consumer-group-ID's be globally unique, or unique per topic? What about consumer-group-names?
- Should "creating a consumer group" be an admin capability, or should any client be able to do it?
- If a single consumer-client is assigned multiple partitions, it seems that the client should be able to parallelize the traffic to its best capability.
  - We can utilize a good async executor to run streaming tasks in parallel (or at least concurrently)
  - Should as client-consumer open a new TCP connection for each partition and stream data in parallel?
  - Should the traffic be multiplexed on a single TCP connection?

### API Considerations

After discussing the initial requirements, we should think about the specific requirements our client API should have. I have a few questions and considerations I have already started thinking about, so I will put them here as a starting point.

- What is the simplest use-case of consumer groups, and how can we make the API to achieve it as lightweight as possible?
  - Perhaps: "I want to consume from group X and do something with each message"

```rust
let mut consumer_group_stream = fluvio.consume_group("topicX", "groupY").stream();
while let Ok(record) = consumer_group_stream.next().await {
    // Do something with 'record'
    // Perhaps 'record' carries information about the offset, consumer group, etc.
    println!("Consumer group: {}", record.consumer_group);
    println!("Record offset: {}", record.offset);
}
```

- How does the client-consumer indicate that a record should be committed?
  - Perhaps when a `Record` is dropped, send a signal to commit its offset?
  - Present the consumer with an explicit API to commit a record?
  - Is the consumer required to commit records in order? (It seems that this should be yes)

```rust
let mut consumer_group_stream = fluvio.consume_group("topicX", "groupY").stream();
while let Ok(record) = consumer_group_stream.next().await {
    // Potential explicit record-commit API
    record.commit();

    // Or, when `record` goes out-of-scope, commit it automatically
}
```

- Should there be a means to _not_ commit a message?
  - What if we want to write the record to a file and we get an IO error?
  - If we use the record-drop-commit strategy, there would need to be some way to abort committing a record

```rust
async fn do_consume_group(fluvio: &Fluvio) -> Result<(), ?> {
    let mut consumer_group_stream = fluvio.consume_group("topicX", "groupY").stream();
    while let Ok(record) = consumer_group_stream.next().await {
        if let Err(e) = std::fs::write("./file.txt", record.as_bytes()) {
            // Mark the record not to be committed
            record.abort_commit();
            return Err(e);
        }
        // If no error, carry on
    }
}
```

---

That's all the thoughts I have in my head right now, comments, critique, and feedback are welcome!

**Comment by saeidakbari** [comment]
This is a beneficial feature as many applications on production with high throughput require multi-instance availability and performance guarantee not to make a single point of failure. In the current Fluvio design consuming can become an easy bottleneck even with smart balancing load to different worker instances, one partition can get hot and eat a lot of resources on a single worker instance.

Also @nicholastmosher , as I reading you ideas, some that I can add:


> * A partition must not be assigned to more than one consumer-client

No, reading from the entire topic no matter if it is partitioned or not has to be load balanced ( like round robin ) between all consumer groups.

> * A consumer group must maintain the currently-committed offset of each partition

This one teoratically brings much more effort in terms of implementation as it requires all `commit` and `acknowledge` processes.




> * Can we assign human-readable names to consumer groups?
> * Should consumer-group-ID's be globally unique, or unique per topic?

Humans are usually error-prone so a systematically auto-generated id for internal communication between Fluvio and the consumer is more likely to be stable. it can be a hybrid solution with a globally unique identifier - like `UUID - with a human-readable name.



> * Should "creating a consumer group" be an admin capability, or should any client be able to do it?

Usually, in message brokers, it is the client's responsibility to create and maintain groups. groups with no member for a while are considered dead and have to be deleted.

**Comment by ruseinov** [comment]
Is there any reason why this or something similar haven't been implemented yet? Right now it's painful to manage, as for each introduced partition one has to spin up exactly one consumer. That presents challenges to programmers and devops alike, as the infrastructure and code then need to be tailored to that nuance. 

I'm guessing [MultiplePartitionConsumer](https://docs.rs/fluvio/0.26.1/fluvio/consumer/struct.MultiplePartitionConsumer.html) wasn't optimal for some reason. Curious to know if there's anything we can do to prioritise or help implementing a way to consume from multiple partitions at a time. 

We're looking to use `Fluvio` in a production system, however this limitation is difficult for us to ignore.

**Comment by drc-infinyon** [comment]
> Is there any reason why this or something similar haven't been implemented yet? Right now it's painful to manage, as for each introduced partition one has to spin up exactly one consumer. That presents challenges to programmers and devops alike, as the infrastructure and code then need to be tailored to that nuance.
> 
> I'm guessing [MultiplePartitionConsumer](https://docs.rs/fluvio/0.26.1/fluvio/consumer/struct.MultiplePartitionConsumer.html) wasn't optimal for some reason. Curious to know if there's anything we can do to prioritise or help implementing a way to consume from multiple partitions at a time.
> 
> We're looking to use `Fluvio` in a production system, however this limitation is difficult for us to ignore.

Thanks for the comment. This is on our road mp. @sehz can provide more details on this.

**Comment by sehz** [comment]
Hi,  `ConsumeGroup` is our roadmap.  I agree, it's important capability in order to scale up consumers.   What's timeline are you looking for?

**Comment by ruseinov** [comment]
> Hi, `ConsumeGroup` is our roadmap. I agree, it's important capability in order to scale up consumers. What's timeline are you looking for?

For us it's basically the sooner the better, as we're now busy converting our MVP to use messaging for communication and scheduling tasks and it's going to need a fair amount of parallelism. The plan is to integrate this week and perhaps skip the parts that need to be more parallel rather than less until there is suitable abstraction in place. 

Within a month or less this is going to be deployed to an actual production system that moves assets back and forth.
